src/premiumvideodevice/PremiumVideoStreamHandler.ts

```diff
@@ -0,0 +1,211 @@
+   */
+  process = async (_event: Event): Promise<void> => {
+    if (!this.inputVideoStream) {
+      return;
+    }
+    const processVideoStart = performance.now();
+
+    // Draw the videoInput onto our input canvas
+    if (this.videoInput.videoWidth) {
+      // Match input canvas to same dimensions as our input stream
+      if (
+        this.canvasInput.width !== this.videoInput.videoWidth ||
+        this.canvasInput.height !== this.videoInput.videoHeight
+      ) {
+        this.canvasInput.width = this.videoInput.videoWidth;
+        this.canvasInput.height = this.videoInput.videoHeight;
+      }
+      this.inputCtx.drawImage(this.videoInput, 0, 0);
+    }
+
+    // Collect out input frame data
+    let inputImageData = this.inputCtx.getImageData(
+      0,
+      0,
+      this.canvasInput.width,
+      this.canvasInput.height
+    );
+
+    // Transform input data
+    let transformedImageData = await this.videoEffectDriver.apply(inputImageData);
+
+    // Confirm that the output canvas is still matching video frame size
+    if (
+      this.canvasOutput.width !== this.videoInput.videoWidth &&
+      this.canvasOutput.height !== this.videoInput.videoHeight
+    ) {
+      this.canvasOutput.width = this.videoInput.videoWidth;
+      this.canvasOutput.height = this.videoInput.videoHeight;
+    }
+
+    // Place transformed data onto output canvas
+    this.outputCtx.putImageData(transformedImageData, 0, 0);
+
+    // timing to maintain requested fps
+    const processVideoLatency = performance.now() - processVideoStart;
+    //const leave = (1000 * 2) / this.framerate - processVideoLatency; // half fps
+    const nextFrameDelay = Math.max(0, 1000 / this.framerate - processVideoLatency);
+
+    // TODO: use requestAnimationFrame which is more organic and allows browser to conserve resources by its choices.
+    /* @ts-ignore */
+    this.lastTimeout = setTimeout(this.process, nextFrameDelay);
+  };
+
+  /**
+   * Stop invoking our transformation function and stop the video tracks associated
+   * with our output media stream
+   */
+  stop() {
+    this.videoInput.removeEventListener('loadedmetadata', this.process);
+    this.videoInput.srcObject = null;
+
+    this.destroyInputMediaAndBuffers();
+
+    // Stop all the output tracks, but don't discard the media stream,
+    // because it's how other parts of the codebase recognize when
+    // a selected stream is part of this transform device.
+    if (this.outputMediaStream) {
+      for (const track of this.outputMediaStream.getVideoTracks()) {
+        track.stop();
+      }
+    }
+
+    // This will prevent the process loop from continuing to call itself
+    if (this.lastTimeOut) {
+      clearTimeout(this.lastTimeOut);
+      this.lastTimeOut = undefined;
+    }
+  }
+
+  /**
+   * Stop the video tracks associated with our input media stream and remove
+   * reference to input source
+   */
+  private destroyInputMediaAndBuffers() {
+    if (this.inputVideoStream) {
```

Comment:
all of this can be `this.inputVideoStream?.getTracks().map(track => track.stop());`

Reply:
Fixed